### PROJET DE BASE DE DONNEES
### Réalisé GIRINEZA BERALDIN

_Sur l'analyse du marché automobile d'occasion dans la métropole Lilloise_

### Objectifs :
- Concevoir un modèle de données en étoile
- Mettre en oeuvre une architecture "Lakehouse" simplifiée sur Databricks
- Pratiquer le langage SQL pour comprendre et transformer les données
- Connecter un outil de Dataviz (PowerBI) à une source Cloud. ![]()


#### PHASE A : Ingestion & analyse exploratoire

Dans cette phase, nous avons importé les données manuellement sur Databricks avant de pourvoir effectuer le traitement et l'analyse. 

In [0]:
# Imporation et description des données avec PySpark

from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("Filter app").getOrCreate()
# 1. Chargement des données
df = spark.read.table("workspace.default.leboncoin_vehicules_lille")

# Exploration des données
display(df.describe())
# Identification des valeurs nulles/abberantes
# count : Le nombre de valeurs non nulles
# mean : Moyenne pour les variables numériques
# stddev: Ecart-type pour les valeurs bumériques
# min : Mininimum des variables numériques
# max : Maximum des valeurs numériques

summary,ID,URL,Titre,Prix,Ville,CP,Vendeur,Description,profile_picture_url,brand,model,u_car_brand,u_car_model,regdate,mileage,fuel,gearbox,doors,seats,issuance_date,vehicle_type,vehicule_color,horsepower,horse_power_din,vehicle_vsp,vehicle_available_payment_methods,vehicle_is_eligible_p2p,car_price_min,car_price_max,car_price_positioning,car_rotation_delay,argus_object_id,vehicle_damage,vehicle_upholstery,vehicle_specifications,vehicle_interior_specs,vehicle_euro_emissions_standard,critair,activity_sector,vehicle_technical_inspection_a,spare_parts_availability,u_car_finition,u_car_version,rating_score,rating_count,ad_warranty_type,vehicle_history_report_public_url,store_logo,store_name,online_store_id,custom_ref,old_price,vehicle_history_report_status,car_contract,monthly_payment_price,vehicle_battery_state_of_health
count,740,741,741,739,740,739,739,740,255,740,740,740,740,740,740,740,740,655,668,712,660,645,632,644,718,257,258,451,452,426,57,591,327,0,0,1,85,352,482,188,482,371,379,94,94,115,243,245,245,245,288,211,209,141,4,1
mean,3.0806020892851353E9,3490.0,232.5,5040.2936400541275,null,59414.72530446549,null,null,null,null,621.8562091503268,2011.0,199000.0,2007.841891891892,155625.33378378378,1.6094594594594596,1.1216216216216217,9164.746564885496,34435.69760479042,null,6.0,109.0,5.699367088607595,96.9860248447205,null,null,24.434108527131784,5473.80266075388,6036.3606194690265,974525.0,0.0,1119982.3637901861,null,null,null,3.0,1.0,1.875,1.0,2026.8670212765958,10.115384615384615,null,null,0.9846808510638299,13.75531914893617,null,null,null,null,3643521.47755102,3.0716758272498E12,6335.682464454976,null,1.0921985815602837,127.75,100.0
stddev,3.022604106306122E7,null,49.66890375275057,1989.2645218215537,null,642.4577524907834,null,null,null,null,1059.65868026542,null,null,73.99265316043453,61866.16490412119,0.730297160663971,0.351016413292335,95342.3966037652,182469.43462132697,null,null,null,4.987604245753571,30.765375961395893,null,null,360.34514032388273,1987.0035527770099,2212.381475256907,null,0.0,341874.60030238656,null,null,null,null,null,1.1278651391366665,0.0,0.411567577929361,9.811531676866455,null,null,0.04144440835883819,31.376892452959336,null,null,null,null,2864117.5904878993,4.2340088614577445E13,1956.440995818584,null,0.2903374691595514,1.5,null
min,2712561101,3490,2025-11-23 07:31:05,1,Allennes-les-Marais,59000,private,"!!!!!!!!!!!! ATTENTION!!!!!!!!!!!!!!!!!BIEN LIRE ANNONCE bsr ;;; vds AUDI a 4 break s line, am 2007 , 204000 km, distribution faite y a 10000 km pneus neufs, beaucoup des factures , ct daté 12/06/2025, ( moins de 6 mois) (ATTENTION !!!!!! ct se termine 12/12/2025 DEPECHEZ VOUS) double clés , options système navigation climatisation auto, jantes alliage, intérieure propre etc. venir voir, photos supplémentaire par SMS prix 3500 € roule parfaitement, idéal pour faire la route tout les renseignement UNIQUEMENT PAR TELEPHONE !!!!!!!!!!!!!!!",Classe A,Alfa Romeo,1007,2011,199000,2,2,1,1,2,1,01/2004,4x4,109,1,4,avecpermis,credit_card,2,1220,0,974525,0,4204,excellent_condition,null,null,3,1,0,1,2025,10,150ème Anniversaire,1.2 16v 75ch Modus.com,0.75,1,essential,https://autoviza.fr/report/report?uid=0120d074-da3d-4d83-b161-8f9ada87b909,https://img.leboncoin.fr/api/v1/lbcpb1/images/01/80/07/0180075060.jpg?rule=bo-thumb,1000 AUTOS,2357,0002,600,broken,1,127,100
max,3099030706,https://www.leboncoin.fr/ad/voitures/3099030706,"🟢 Ssangyong Tivoli 1.6 e-XGi 2WD 128ch Urban ""1ère Main-Crit’Air 1""",8000,pro,62840,pro,"🚘 Clio 3 – Diesel – 2010 – 156 000 km – Toit ouvrant ! Superbe Renault Clio 3, fiable, économique et agréable à conduire ! Toujours bien entretenue, parfaite pour un premier véhicule ou pour les trajets quotidiens. ✅ Année : 2010 ✅ Kilométrage : 156 000 km ✅ Motorisation : Diesel – faible consommation ✅ Équipements : Toit ouvrant, climatisation, vitres électriques, direction assistée ✅ Contrôle technique : OK – rien à prévoir ⚠️ Petit coup sur le pare-chocs avant (rien de grave, uniquement e

A partir de cette description on se rend compte que la base de données importée est constituée de 740 observations/clients dont un n'ayant pas d'identifiant. On remarque par ailleurs la présente des valeurs null sur une ligne des deux lignes des données, du fait que les attributs 'Prix', 'CP', 'Vendeur' n'ont que 739 valeurs non nulles. Aussi, le prix minimum est de 1 ce qui est anormale, nous remplacçons alors par la valeur moyenne

In [0]:
# Analyse de la structure des données
df.printSchema()

root
 |-- ID: long (nullable = true)
 |-- URL: string (nullable = true)
 |-- Titre: string (nullable = true)
 |-- Prix: long (nullable = true)
 |-- Date: timestamp (nullable = true)
 |-- Ville: string (nullable = true)
 |-- CP: long (nullable = true)
 |-- Vendeur: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- profile_picture_url: string (nullable = true)
 |-- brand: string (nullable = true)
 |-- model: string (nullable = true)
 |-- u_car_brand: string (nullable = true)
 |-- u_car_model: string (nullable = true)
 |-- regdate: long (nullable = true)
 |-- mileage: long (nullable = true)
 |-- fuel: long (nullable = true)
 |-- gearbox: long (nullable = true)
 |-- doors: long (nullable = true)
 |-- seats: long (nullable = true)
 |-- issuance_date: string (nullable = true)
 |-- vehicle_type: string (nullable = true)
 |-- vehicule_color: string (nullable = true)
 |-- horsepower: long (nullable = true)
 |-- horse_power_din: long (nullable = true)
 |-- vehicle_vsp: str

In [0]:
%sql
/* Création d'une copie de la table brute */

CREATE OR REPLACE TABLE bronze_annonces AS
SELECT *
FROM leboncoin_vehicules_lille;

/* Correction des données pour lesquelles le prix est égal à 1 */

UPDATE bronze_annonces
SET prix = (
    SELECT AVG(prix)
    FROM bronze_annonces
    WHERE prix <> 1
      AND prix IS NOT NULL
)
WHERE prix = 1;


num_affected_rows
2


#### PHASE B : Modélisation (Conception)

Dans l’architecture Delta Lake (Databricks), on organise les données en trois niveaux : 
- Bronze : Données brutes directement importées
- Silver : Données nettoyées, normalisées
- Golden : Données prêtes pour l'analyse ou le reporting

Nous vons donc fait une modélisation en étoile, en créant une table de fait et 4 tables de dimensions (date, vehicule, localisation et vendeur) lesquelles ont été alimentées à partir de la table silver. 

Le modèle en étoile que nous avons mis en place permet de structurer les données de manière simple, lisible et optimisée pour faciliter l'exploitation et l’analyse. La séparation entre la table de faits (contenant les mesures comme le prix ou le kilométrage) et 4 tables de dimensions (véhicule, date, localisation, vendeur) facilite les jointures, améliore les performances des requêtes analytiques et rend le modèle compatible avec les outils de reporting, Power BI en particulier. 


In [0]:
%sql
/* TRAITEMENT DES DONNEESS : Création de la table de silver */

CREATE OR REPLACE VIEW silver_annonces AS
SELECT
  CAST(ID AS BIGINT) AS id_annonce,
  -- Dates
  TO_DATE(Date, 'yyyy-MM-dd') AS date_annonce,
  
  -- Prix & kilométrage
  CAST(prix AS INT) AS prix,
  CAST(mileage AS INT) AS kilometrage,
  
  -- Motorisation
  CASE fuel
    WHEN '1' THEN 'Essence'
    WHEN '2' THEN 'Diesel'
    WHEN '4' THEN 'Electrique'
    ELSE 'Inconnu'
  END AS motorisation,
  
  -- Boîte de vitesse
  CASE gearbox
    WHEN '1' THEN 'Manuelle'
    WHEN '2' THEN 'Automatique'
    ELSE 'Inconnu'
  END AS type_boite,
  
  -- Vendeur
  CASE Vendeur
    WHEN 'private' THEN 'Particulier'
    WHEN 'pro' THEN 'Professionnel'
    ELSE 'Inconnu'
  END AS type_vendeur,
  
  -- Géographie
  regexp_extract(cp, '(\\d{5})', 1) AS code_postal,
  TRIM(ville) AS ville,
  
  -- Autres colonnes
  brand,
  model,
  regdate,
  nvl(vehicle_type, 'Autre') AS type_vehicule
FROM bronze_annonces
WHERE Prix IS NOT NULL;

In [0]:
%sql
/* Visualisation de la table silver*/
select * from silver_annonces

id_annonce,date_annonce,prix,kilometrage,motorisation,type_boite,type_vendeur,code_postal,ville,brand,model,regdate,type_vehicule
3099030706,2025-11-22,5300,250000,Diesel,Manuelle,Particulier,59200,Tourcoing,Nissan,Autres,2010,4x4
3099027749,2025-11-22,4500,216000,Diesel,Manuelle,Particulier,59370,Mons-en-Baroeul,Kia,Autres,2012,break
3099024200,2025-11-22,7800,32000,Electrique,Automatique,Particulier,59150,Wattrelos,Dacia,Autres,2022,4x4
3098859515,2025-11-22,7990,123000,Diesel,Manuelle,Professionnel,59155,Faches-Thumesnil,Bmw,Serie 1,2012,berline
3099019487,2025-11-22,5950,214500,Diesel,Manuelle,Particulier,59200,Tourcoing,Opel,Cascada,2014,cabriolet
3099016758,2025-11-22,6300,278000,Inconnu,Automatique,Particulier,59000,Lille,Toyota,Prius,2010,berline
3099016003,2025-11-22,6700,131000,Essence,Manuelle,Particulier,59000,Lille,Peugeot,308,2015,berline
3083614171,2025-11-22,8000,168449,Diesel,Manuelle,Professionnel,59136,Wavrin,Land Rover,Autres,2016,4x4
3099007955,2025-11-22,3800,200000,Diesel,Manuelle,Particulier,59120,Loos,Citroen,C3,2012,berline
3088762385,2025-11-22,5990,86870,Essence,Manuelle,Professionnel,59930,La Chapelle-d'Armentières,Mazda,2,2010,berline


In [0]:
%sql
/* CREATION DES TABLES DE DIMENSIONS : VEHICULE */
CREATE OR REPLACE TABLE dim_vehicule AS
SELECT
  CONCAT('veh', ROW_NUMBER() OVER (
      ORDER BY marque, modele, motorisation, type_vehicule, type_boite
  )) AS id_vehicule,
  marque,
  modele,
  motorisation,
  type_vehicule,
  type_boite
FROM (
  SELECT DISTINCT
    brand AS marque,
    model AS modele,
    motorisation,
    type_vehicule,
    type_boite
  FROM silver_annonces
) t;




num_affected_rows,num_inserted_rows


In [0]:
%sql
/* CREATION DES TABLES DE DIMENSIONS : DATE */
CREATE OR REPLACE TABLE dim_date AS
SELECT
  CONCAT('dat', ROW_NUMBER() OVER (ORDER BY date_annonce)) AS id_date,
  date_annonce AS date,
  YEAR(date_annonce) AS annee,
  MONTH(date_annonce) AS mois,
  DAY(date_annonce) AS jour,
  DATE_FORMAT(date_annonce, 'EEEE') AS jour_semaine
FROM (
  SELECT DISTINCT
    date_annonce
  FROM silver_annonces
  WHERE date_annonce IS NOT NULL
) d;


num_affected_rows,num_inserted_rows


In [0]:
%sql
/* CREATION DES TABLES DE DIMENSIONS : LOCALISATION */

CREATE OR REPLACE TABLE dim_localisation AS
SELECT
  CONCAT('loc', ROW_NUMBER() OVER (ORDER BY code_postal, ville)) AS id_localisation,
  code_postal,
  ville
FROM (
  SELECT DISTINCT
    code_postal,
    ville
  FROM silver_annonces
  WHERE code_postal IS NOT NULL AND ville IS NOT NULL
) l;



num_affected_rows,num_inserted_rows


In [0]:
%sql
/* CREATION DES TABLES DE DIMENSIONS : VENDEUR */

CREATE OR REPLACE TABLE dim_vendeur AS
SELECT
  CONCAT('ven', ROW_NUMBER() OVER (ORDER BY type_vendeur)) AS id_vendeur,
  type_vendeur
FROM (
  SELECT DISTINCT
    type_vendeur
  FROM silver_annonces
  WHERE type_vendeur IS NOT NULL
) v;



num_affected_rows,num_inserted_rows


In [0]:
%sql
/* CREATION DE LA TABLE DE FAIT */

CREATE OR REPLACE TABLE fact_annonce AS
SELECT DISTINCT
    s.id_annonce,
    d.id_date,
    v.id_vehicule,
    l.id_localisation,
    ven.id_vendeur,
    s.prix,
    s.kilometrage
FROM silver_annonces s
LEFT JOIN dim_date d
    ON s.date_annonce = d.date
LEFT JOIN dim_vehicule v
    ON s.brand          = v.marque
   AND s.model          = v.modele
   AND s.motorisation   = v.motorisation
   AND s.type_vehicule  = v.type_vehicule
   AND s.type_boite     = v.type_boite
LEFT JOIN dim_localisation l
    ON s.code_postal = l.code_postal
   AND s.ville       = l.ville
LEFT JOIN dim_vendeur ven
    ON s.type_vendeur = ven.type_vendeur
WHERE s.id_annonce IS NOT NULL;


num_affected_rows,num_inserted_rows


Après avoir chargé les différentes tables, nous remarquons que la table de fait comporte 738 observatios au lieu de 739 que comporte la table silver. Pour identifier le problème, nous avons recherché les identifiants des annonces en double. Nous avons ainsi consaté qu'un identifiant de l'annonce était en double, et ces doublants comporte les même informations sur tous les attributs de la base de donnée. Donc la table de fait est bien correcte avec 738 observations

In [0]:
%sql
/* Détection des doublons sur les identifiants */

SELECT id_annonce, COUNT(*) as valeur_double
FROM silver_annonces
GROUP BY id_annonce
HAVING COUNT(*) > 1;


id_annonce,valeur_double
3098260184,2


#### Vérification des prix dérisoirs

En faisant une analyse descriptive du montant des véhicules, on remarque une plage des prix allant de 169 à 8 000 avec une moyenne de 5 044.96. Sur cette base, il se peut qu'il y ait des prix erronés. Nous faisons une vérification sur un seuil de 500. 

In [0]:
%sql
SELECT min(f.prix) AS prix_min, round(avg(f.prix),2) AS prix_moyen, max(f.prix) AS prix_max from fact_annonce f 

prix_min,prix_moyen,prix_max
169,5058.65,8000


In [0]:
%sql
/* Vérification des prix dérisoires */

SELECT *
FROM fact_annonce f
JOIN dim_vehicule v ON f.id_vehicule = v.id_vehicule
WHERE f.prix <= 500; 


id_annonce,id_date,id_vehicule,id_localisation,id_vendeur,prix,kilometrage,id_vehicule,marque,modele,motorisation,type_vehicule,type_boite
3098859256,dat2,veh53,loc53,ven1,169,235850,veh53,Citroen,C3,Essence,berline,Manuelle
3098587823,dat2,veh68,loc42,ven1,500,216000,veh68,Citroen,C4 Picasso,Diesel,autre,Automatique
3024330452,dat1,veh137,loc69,ven1,300,254396,veh137,Hyundai,Matrix,Diesel,autre,Manuelle
3098822489,dat2,veh209,loc22,ven1,300,260000,veh209,Opel,Corsa,Diesel,berline,Manuelle
3097969612,dat1,veh312,loc1,ven1,500,168000,veh312,Renault,Clio,Essence,berline,Manuelle
3098340965,dat1,veh401,loc27,ven1,379,20500,veh401,Volkswagen,Golf,Essence,berline,Manuelle
3098601265,dat2,veh312,loc26,ven1,300,225000,veh312,Renault,Clio,Essence,berline,Manuelle
